### ISSUE TITLE
Make pipelines more fault-tolerant

### ISSUE THREAD

**Issue by derekchiang** [pachyderm/pachyderm#300]
Currently, the logic for running a pipeline resides in the `runPipeline` function in `src/server/pps/server/api_server.go`.  The function can however return prematurely due to transient errors, and we don't have any logic for rerunning the pipeline when that happens.

For instance, if `runPipeline` is blocked on `ListCommit`, and `ListCommit` fails because a pfs node goes down, `runPipeline` will return with an error.

Basically, we need to distinguish between unrecoverable and transient errors, and rerun a pipeline if it failed due to transient errors.

**Comment by jdoliner** [comment]
What's an example of an unrecoverable error for which we should stop the pipeline completely?
It might well be that what we want to do here is just log errors that `runPipeline` gets but always restart it. I haven't thought about it too much though so that might not actually be the right idea.

**Comment by derekchiang** [comment]
If the pps node that's running a pipeline experiences a network partition from the pfs node that's pulsing the next input commit, the pps node will be unable to run the pipeline until the network partition repairs.  If the partition is persistent, the pipeline will never get to be run.

So, what do you think of an exponential backoff mechanism, plus the policy that if we have backed off for more than n times (where n = 3, for instance), we crash the node by panic() in order to force the pipeline to be reassigned to some other node?

Though, by crashing the node, we might experience the cascading failure scenario that you described offline.

It seems like we can't have a perfect solution unless we can separate pps and pfs.  What do you think we should do for 1.0?  Probably simply have exponential backoff and hope that all errors will be transient?

**Comment by derekchiang** [comment]
Closing this issue due to the fix in #306.  The rest of the solution will be provided with #305.
